swarm/network/simulations/discovery/dynamic_test.go

```diff
@@ -0,0 +1,359 @@
+	adapter := paramstring[3]
+
+	bootNodes = make([]*discover.NodeID, 3)
+	events = make(chan *p2p.PeerEvent)
+	subs = make(map[discover.NodeID]*rpc.ClientSubscription)
+	rpcs = make(map[discover.NodeID]*rpc.Client)
+	ids = make([]discover.NodeID, nodeCount)
+
+	log.Info("dynamic test", "nodecount", nodeCount, "adaptertype", adapter)
+
+	var a adapters.NodeAdapter
+	if adapter == "exec" {
+		dirname, err := ioutil.TempDir(".", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+		a = adapters.NewExecAdapter(dirname)
+	} else if adapter == "sock" {
+		a = adapters.NewSocketAdapter(services)
+	} else if adapter == "tcp" {
+		a = adapters.NewTCPAdapter(services)
+	} else if adapter == "sim" {
+		a = adapters.NewSimAdapter(services)
+	}
+	// create network
+	net := simulations.NewNetwork(a, &simulations.NetworkConfig{
+		ID:             "0",
+		DefaultService: serviceName,
+	})
+	defer net.Shutdown()
+
+	for i := 0; i < int(nodeCount); i++ {
+		conf := adapters.RandomNodeConfig()
+		node, err := net.NewNodeWithConfig(conf)
+		if err != nil {
+			t.Fatalf("error starting node: %s", err)
+		}
+		ids[i] = node.ID()
+		log.Info("new node", "id", ids[i])
+		if i < bootNodeCount {
+			bootNodes[i] = &ids[i]
+		}
+
+		if err = net.Start(node.ID()); err != nil {
+			t.Fatalf("error starting node %s: %s", node.ID().TerminalString(), err)
+		}
+		client, err := node.Client()
+		if err != nil {
+			t.Fatal(err)
+		}
+		rpcs[ids[i]] = client
+		sub, err := client.Subscribe(context.Background(), "admin", events, "peerEvents")
+		if err != nil {
+			t.Fatalf("error getting peer events for node %v: %s", ids[i], err)
+		}
+		subs[ids[i]] = sub
+	}
+
+	ctrl := newNodeCtrl(net, ids)
+
+	trigger := make(chan discover.NodeID)
+	// run a simulation which connects the nodes to the bootnodes
+	action := func(ctx context.Context) error {
+		for i := 0; i < int(nodeCount); i++ {
+			var j int
+			if i < bootNodeCount {
+				if i == bootNodeCount-1 {
+					j = 0
+				} else {
+					j = i + 1
+				}
+			} else {
+				j = rand.Intn(bootNodeCount)
+			}
+			go func(i, j int) {
+				net.Connect(ids[i], ids[j])
+				// TODO: replace with simevents check
+				time.Sleep(time.Second * 2)
+				trigger <- ids[i]
+			}(i, j)
+		}
+		return nil
+	}
+
+	// construct the peer pot, so that kademlia health can be checked
+	check := func(ctx context.Context, id discover.NodeID) (bool, error) {
+		log.Warn("checking", "id", id)
+		var err error
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		case e := <-events:
+			if e.Type == p2p.PeerEventTypeAdd {
+				log.Info("check got add", "id", id)
+				n := net.GetNode(id)
+				err := ctrl.checkHealth(n)
+				log.Debug("check health return", "id", id, "err", err)
+				return true, err
+			}
+		case err := <-subs[id].Err():
+			if err != nil {
+				log.Error(fmt.Sprintf("error getting peer events for node %v", id), "err", err)
+			}
+		default:
+		}
+		return false, err
+	}
+
+	timeout := 30 * time.Second
```

Comment:
What's the planned replacement for this timeout

Reply:
None, this is the expiry for the simulation step.

Reply:
Oh...haha sorry, I fantasized this timeout to be a Sleep....

Reply:
I'm lazy, but not _that_ lazy